## Project Name : Text Characterization using Speech Recognition

#### In this project we will see how we can use Speech Recognition librarties to characterize audio speech files by importing or by just saying them. We will see how we can perform different operations on a speech to text converted data.

### Time Line of the project :
- Importing Libraries
- Using Speech Recognition
- Observing the audio data
- Analyzing the lyrics of a song

In [ ]:
#!pip install SpeechRecognition

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 32.8/32.8 MB 34.3 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import numpy as np
import speech_recognition as sr

In [ ]:
import IPython.display as ipd

In [ ]:
music = ipd.Audio('/content/drive/MyDrive/Deeep Learning Project/DL Projects/Text Characterization using Speech Recognition/Code and Files/voice-data.wav')

In [ ]:
music

Output hidden; open in https://colab.research.google.com to view.

In [ ]:
recognizer = sr.Recognizer()
audio_file = sr.AudioFile('/content/drive/MyDrive/Deeep Learning Project/DL Projects/Text Characterization using Speech Recognition/Code and Files/voice-data.wav')
type(audio_file)

speech_recognition.AudioFile

**APIs : There are various APIs present which does the Speech To Text Conversion. Different APIs are :**

- IBM Speech TO Text
- CMU SPhinx
- Google Web Speech API (Recognize Google)
- Google Cloud Speech
And more

**Google Web Speech API:**

This is a free service provided by Google i.e. Recognize Google, through which we can convert audio speech files to text and perform operations on them

In [ ]:
with sr.AudioFile('/content/drive/MyDrive/Deeep Learning Project/DL Projects/Text Characterization using Speech Recognition/Code and Files/voice-data.wav') as source:
    audio_file = recognizer.record(source)
    result=recognizer.recognize_google(audio_data=audio_file)

In [ ]:
music

Output hidden; open in https://colab.research.google.com to view.

In [ ]:
result

"Eden would you be then you know locking this meeting in so that we don't have no participants because once the division of team is done and stuff like that what does it or is it okay if people if more people join in the game you can play I think we have about 18 people like that most people would know about it so we playing code names guys it's a fun board game just just enjoy yourself have people join in case they want to do and then once we started okay so let's get started my screen guys like this is the time for rules if you don't understand now so the game that you're going to play is called code names and in the game you're all going to be like spy agents so much and the other side of team will look at it and before I explained for the how many of you know the game coordinates how many of you play the game for kids before Akash kindly call me if I'm going wrong somewhere else play coordinates it's good for us because we're going to be done came today so basically put up into two

There are two types of taind input:

1)Duration: This is used to select a specific time of audio data i.e. If you wnat to select just 5 seconds then you can set Duration=5 and it will only select the 5 seconds of audio file

In [ ]:
with sr.AudioFile('/content/drive/MyDrive/Deeep Learning Project/DL Projects/Text Characterization using Speech Recognition/Code and Files/voice-data.wav') as source:
    audio_file_duration = recognizer.record(source, duration=10.0)
    result1=recognizer.recognize_google(audio_data=audio_file_duration)

In [ ]:
result1

"Eden awesome okay would you be then you know locking this meeting in so that we don't have"

2)Offset: This is used when you cut out some part of the starting data from your audio file i.e. if you don't want first 2 seconds then you can select offset as 2 and it willl stop the first two seconds.

In [ ]:
with sr.AudioFile('/content/drive/MyDrive/Deeep Learning Project/DL Projects/Text Characterization using Speech Recognition/Code and Files/voice-data.wav') as source:
    audio_file_offset = recognizer.record(source, offset=5.0)
    result2=recognizer.recognize_google(audio_data=audio_file_offset)

In [ ]:
result2

"do you know locking this meeting in so that we don't have no participants because once the division of team is done and stuff like that what does it or is it okay if people if more people join in the game I think we have about 1 8. 7 5 5"

Combining both the file the duration and the offset

In [ ]:
with sr.AudioFile('/content/drive/MyDrive/Deeep Learning Project/DL Projects/Text Characterization using Speech Recognition/Code and Files/voice-data.wav') as source:
    audio_file_combine = recognizer.record(source, duration=10.0, offset=5.0)
    result3=recognizer.recognize_google(audio_data=audio_file_combine)

In [ ]:
result3

"do you know locking this meeting in so that we don't have no participants because once the division of team is done and stuff like that"

Comparing all the result

In [ ]:
print(result)  #full text from the music audio file
print(result1)  #applying the duration as 10 sec
print(result2)  #applying the offset as 5 sec
print(result3)  #Combning both the result1 and result2

Eden would you be then you know locking this meeting in so that we don't have no participants because once the division of team is done and stuff like that what does it or is it okay if people if more people join in the game you can play I think we have about 18 people like that most people would know about it so we playing code names guys it's a fun board game just just enjoy yourself have people join in case they want to do and then once we started okay so let's get started my screen guys like this is the time for rules if you don't understand now so the game that you're going to play is called code names and in the game you're all going to be like spy agents so much and the other side of team will look at it and before I explained for the how many of you know the game coordinates how many of you play the game for kids before Akash kindly call me if I'm going wrong somewhere else play coordinates it's good for us because we're going to be done came today so basically put up into two 

**Effect of Noise**

Noise in the background data can create disturbances or error in the results hence it is necessary to remove the noise from the audio file.

In [ ]:
with sr.AudioFile('/content/drive/MyDrive/Deeep Learning Project/DL Projects/Text Characterization using Speech Recognition/Code and Files/voice-data.wav') as source:
    recognizer.adjust_for_ambient_noise(source, duration=0.5) # fix typo in method name
    audio_file_noise = recognizer.record(source)

result4=recognizer.recognize_google(audio_data=audio_file_noise)

In [ ]:
music

Output hidden; open in https://colab.research.google.com to view.

In [ ]:
result4

"would you be then you know locking this meeting in so that we don't have no participants because once the division of team is done and stuff like that okay or does it or is it okay if people if more people join in the game you can I think we have about 18 people and stuff like that most people would know about it so we playing code names guys it's a fun board game just just enjoy yourself happy okay so let's get started my screen guys like this time for girls to be fine now so the game that you're going to play is called code names and in the game you're all going to be like this so I was up into two different one set of group will belong to the blues party and the other side of team will look at it and before I explained for the how many of you know the game coordinates how many of you play the game for kids before Akash kindly call me if I'm going wrong somewhere good for us because we're going to be done came today so coach up into two different teams Blue Team and Red Team and mos

**Text Analysis**

In [ ]:
result_str=result.split(' ')

In [ ]:
print(result_str)

['Eden', 'would', 'you', 'be', 'then', 'you', 'know', 'locking', 'this', 'meeting', 'in', 'so', 'that', 'we', "don't", 'have', 'no', 'participants', 'because', 'once', 'the', 'division', 'of', 'team', 'is', 'done', 'and', 'stuff', 'like', 'that', 'what', 'does', 'it', 'or', 'is', 'it', 'okay', 'if', 'people', 'if', 'more', 'people', 'join', 'in', 'the', 'game', 'you', 'can', 'play', 'I', 'think', 'we', 'have', 'about', '18', 'people', 'like', 'that', 'most', 'people', 'would', 'know', 'about', 'it', 'so', 'we', 'playing', 'code', 'names', 'guys', "it's", 'a', 'fun', 'board', 'game', 'just', 'just', 'enjoy', 'yourself', 'have', 'people', 'join', 'in', 'case', 'they', 'want', 'to', 'do', 'and', 'then', 'once', 'we', 'started', 'okay', 'so', "let's", 'get', 'started', 'my', 'screen', 'guys', 'like', 'this', 'is', 'the', 'time', 'for', 'rules', 'if', 'you', "don't", 'understand', 'now', 'so', 'the', 'game', 'that', "you're", 'going', 'to', 'play', 'is', 'called', 'code', 'names', 'and', 'i

In [ ]:
result_str[5]

'you'

**Different number of words used?**

In [ ]:
unique_words = set(result_str)
print(unique_words)

{'Akash', 'then', 'kindly', 'here', 'only', 'make', 'about', 'stuff', 'team', 'me', 'good', 'have', 'Team', 'do', 'just', 'teams', 'joining', 'see', 'somewhere', 'what', 'called', "don't", 'because', 'us', 'blue', 'guys', 'or', 'but', 'people', 'be', 'case', 'my', 'from', 'done', 'their', 'has', 'this', 'by', 'words', 'for', 'agents', 'in', 'can', 'word', 'red', "I'll", 'names', "I'm", 'Master', 'knows', 'time', 'is', 'fun', 'to', 'locking', 'rules', 'join', 'nine', 'they', 'think', 'much', 'put', 'many', 'most', 'where', 'know', 'the', 'aim', 'it', 'hidden', 'side', 'coordinates', 'operative', 'both', 'example', '2500', 'protected', 'spider', 'screen', 'once', '18', 'Red', 'look', 'how', 'you', 'and', 'operators', 'into', 'like', 'that', 'at', 'wrong', 'would', 'game', 'eight', 'money', 'playing', 'we', 'meeting', 'among', 'two', 'some', 'respective', 'today', 'Eden', 'explained', 'started', 'will', 'board', 'agency', 'enjoy', 'came', 'location', 'Blue', 'get', 'if', 'spy', 'I', 'does

In [ ]:
print("The number of different words used: ",len(unique_words))

The number of different words used:  153


**Count the repetition of the words**

First we will store the unique words in a dictionary

In [ ]:
#To count the number of times the unique words apper, first in the unique_word lust
word_dict = {} #An empth dictionary
for word in result_str:
    word_dict[word] = 0
print(word_dict)

{'Eden': 0, 'would': 0, 'you': 0, 'be': 0, 'then': 0, 'know': 0, 'locking': 0, 'this': 0, 'meeting': 0, 'in': 0, 'so': 0, 'that': 0, 'we': 0, "don't": 0, 'have': 0, 'no': 0, 'participants': 0, 'because': 0, 'once': 0, 'the': 0, 'division': 0, 'of': 0, 'team': 0, 'is': 0, 'done': 0, 'and': 0, 'stuff': 0, 'like': 0, 'what': 0, 'does': 0, 'it': 0, 'or': 0, 'okay': 0, 'if': 0, 'people': 0, 'more': 0, 'join': 0, 'game': 0, 'can': 0, 'play': 0, 'I': 0, 'think': 0, 'about': 0, '18': 0, 'most': 0, 'playing': 0, 'code': 0, 'names': 0, 'guys': 0, "it's": 0, 'a': 0, 'fun': 0, 'board': 0, 'just': 0, 'enjoy': 0, 'yourself': 0, 'case': 0, 'they': 0, 'want': 0, 'to': 0, 'do': 0, 'started': 0, "let's": 0, 'get': 0, 'my': 0, 'screen': 0, 'time': 0, 'for': 0, 'rules': 0, 'understand': 0, 'now': 0, "you're": 0, 'going': 0, 'called': 0, 'all': 0, 'spy': 0, 'agents': 0, 'much': 0, 'other': 0, 'side': 0, 'will': 0, 'look': 0, 'at': 0, 'before': 0, 'explained': 0, 'how': 0, 'many': 0, 'coordinates': 0, 'kids

In [ ]:
for word in result_str:
    word_dict[word] = word_dict[word] + 1
print('The count for each word spoken number of times are: ',word_dict)

The count for each word spoken number of times are:  {'Eden': 1, 'would': 2, 'you': 10, 'be': 6, 'then': 3, 'know': 3, 'locking': 1, 'this': 4, 'meeting': 1, 'in': 6, 'so': 11, 'that': 4, 'we': 4, "don't": 2, 'have': 4, 'no': 1, 'participants': 1, 'because': 2, 'once': 2, 'the': 17, 'division': 1, 'of': 8, 'team': 9, 'is': 6, 'done': 2, 'and': 13, 'stuff': 1, 'like': 4, 'what': 1, 'does': 1, 'it': 4, 'or': 1, 'okay': 2, 'if': 4, 'people': 5, 'more': 1, 'join': 3, 'game': 6, 'can': 2, 'play': 4, 'I': 3, 'think': 1, 'about': 2, '18': 1, 'most': 2, 'playing': 1, 'code': 2, 'names': 2, 'guys': 2, "it's": 2, 'a': 2, 'fun': 1, 'board': 1, 'just': 2, 'enjoy': 1, 'yourself': 1, 'case': 2, 'they': 1, 'want': 1, 'to': 5, 'do': 1, 'started': 2, "let's": 1, 'get': 1, 'my': 2, 'screen': 1, 'time': 1, 'for': 7, 'rules': 1, 'understand': 1, 'now': 1, "you're": 2, 'going': 4, 'called': 1, 'all': 1, 'spy': 2, 'agents': 3, 'much': 1, 'other': 1, 'side': 1, 'will': 6, 'look': 1, 'at': 1, 'before': 2, 'ex

In [ ]:
cols = ['Repetition']
count_df = pd.DataFrame.from_dict(word_dict, orient='index', columns=cols)

In [ ]:
count_df

Repetition
Eden            1
would           2
you            10
be              6
then            3
...           ...
make            1
some            1
money           1
from            1
I'll            1

[153 rows x 1 columns]

In [ ]:
count_df = count_df.reset_index()

In [ ]:
count_df

index  Repetition
0     Eden           1
1    would           2
2      you          10
3       be           6
4     then           3
..     ...         ...
148   make           1
149   some           1
150  money           1
151   from           1
152   I'll           1

[153 rows x 2 columns]

In [ ]:
count_df = count_df.rename(columns={'index': 'Words'})

In [ ]:
count_df

Words  Repetition
0     Eden           1
1    would           2
2      you          10
3       be           6
4     then           3
..     ...         ...
148   make           1
149   some           1
150  money           1
151   from           1
152   I'll           1

[153 rows x 2 columns]

**Counting the number of words spoken per minute**

In [ ]:
print('Total number of words: ',len(result_str))

Total number of words:  342


In [ ]:
print('Total Length of audio: 3.08 minutes')

Total Length of audio: 3.08 minutes


In [ ]:
print('Total number of words spoken per minute: ',(len(result_str)/3.08))

Total number of words spoken per minute:  111.03896103896103


**Approx total number of words spoken are 111 words per minute**